lib/compatibility.bzl

```diff
@@ -0,0 +1,171 @@
+"""Skylib module of convenience functions for `target_compatible_with`.
+
+Load the macros as follows in your `BUILD` files:
+```python
+load("@bazel_skylib//lib:compatibility.bzl", "compatibility")
+```
+
+See the [Platform docs](https://bazel.build/docs/platforms#skipping-incompatible-targets) for
+more information.
+"""
+
+load(":selects.bzl", "selects")
+
+def _get_name_from_target_list(targets, joiner=" or "):
+    """Join a list of strings into a string which is suitable as a target name.
+
+    Removes/replaces characters which are not valid as target names.
+
+    Args:
+      target_list: A list of target names.
+      joiner: An optional string to use for joining the list.
+
+    Returns:
+      A string which is a valid target name.
+    """
+    targets_name = joiner.join([s.split(":")[1] for s in targets])
+    name = targets_name.replace("//", "").replace(":", "_").replace("/", "_").replace(".", "_")
+    return name
+
+def _maybe_make_unique_incompatible_value(name):
+    """Creates a `native.constraint_value` which is "incompatible."
+
+    When composing selects which could all resolve to "incompatible" we need distinct labels.
```

Comment:
Just out of curiosity: Is there a particular reason for why `target_compatible_with` can't deduplicate it's value before processing it? That won't help simplify this for existing Bazel versions though.

Reply:
I suspect that the reason is that there's some fairly generic logic somewhere that complains about duplicate labels in `label_list` attributes. E.g. I remember that adding duplicate deps to `py_library` also throws a bazel error. In other words, I suspect that this behaviour cannot be changed without affecting the same logic for all other attributes. Unless we hard-code something for `target_compatible_with`.

Reply:
True. It's pretty easy to hit this error when programmatically adding dependencies from a macro.

@gregestren Can you perhaps shed some light on why Bazel errors out rather than stably deduping the label list?

Reply:
@philsc and I followed up on this on another issue/PR (which was that?). I assume the historic logic is to encourage tidy BUILD curation. I ran this by the devs most concerned about API consistency and didn't hear strong opinion on keeping it this way.